polyglot/piranha/visualize_rules_graph.py

```diff
@@ -0,0 +1,234 @@
+
+# REQUIRED: pip install toml
+# REQUIRED: pip install graphviz
+
+
+@dataclass(frozen=True, eq=True)
+class Edge:
+    to: str
+    scope: str
+
+
+def collect_rules_groups_edges():
+    for config_path in args.configurations_path:
+        config_path = os.path.abspath(config_path) + os.path.sep
+        rules_file = config_path + 'rules.toml'
+
+        # rules.toml should exist, deliberately fail otherwise
+        rules_toml_dict = toml.load(rules_file)
+        collect_rules_and_groups(rules_toml_dict)
+
+        # edges are optional for input rules (i.e., non built-in)
+        edges_file = config_path + 'edges.toml'
+        if os.path.exists(edges_file):
+            edges_toml_dict = toml.load(edges_file)
+            collect_edges(edges_toml_dict)
+
+
+def collect_rules_and_groups(rules_toml_dict):
+    """
+    Collects rules and groups to further build graph nodes.
+
+    Rules belonging to a group different than `Cleanup Rule` are displayed below the group name.
+
+    Cleanup Rule's are added to a set as they are displayed differently.
+        They get an additional label `(Cleanup_Rule)`.
+
+    Nodes without a group are added to a different set.
+    """
+    for rule_toml in rules_toml_dict['rules']:
+        rule_name: str = sanitize_name(rule_toml['name'])
+        if 'groups' in rule_toml:
+            rule_groups = rule_toml['groups']
+            collect_node_for_group_rule(rule_name, rule_groups)
+        else:  # rule without group
+            orphan_nodes.add(rule_name)
+
+
+def collect_node_for_group_rule(rule_name: str, rule_groups: 'list[str]'):
+    for group_name in rule_groups:
+        group_name = sanitize_name(group_name)
+        if group_name != 'Cleanup_Rule':
+            if group_name in rules_by_group_dict:
+                rules_by_group_dict[group_name].append(rule_name)
+            else:
+                rules_by_group_dict[group_name] = [rule_name]
+        else:  # we don't want to group 'Cleanup Rule's under the same shape
+            cleanup_rules.add(rule_name)
+
+
+def sanitize_name(s: str) -> str:
+    """Graphviz does not like names with spaces. Converts spaces to '_'."""
+    s = re.sub(r"\s+", '_', s)
+    return s
+
+
+def collect_edges(edges_toml_dict):
+    """
+    Groups outgoing edges by rule/group.
+
+    All names are sanitized to replace empty spaces by `_`.
+    """
+    for edge_toml in edges_toml_dict['edges']:
+        from_node = sanitize_name(edge_toml['from'])
+        to_nodes: 'list[str]' = edge_toml['to']
+        scope = sanitize_name(edge_toml['scope'])
+        for to_node in to_nodes:
+            to_node = sanitize_name(to_node)
+            edge = Edge(to=to_node, scope=scope)
+            if from_node in outgoing_edges_by_node:
+                outgoing_edges_by_node[from_node].append(edge)
+            else:
+                outgoing_edges_by_node[from_node] = [edge]
+
+
+def group_rule_names_label(rule_names: 'list[str]') -> 'list[str]':
```

Comment:
What are we grouping here? 
When you say `group_<something>`  I guess the function should return a `dict` or a `list of list`.

Reply:
I guess this is the _"group"_ from our `Rule`. :|

Reply:
I changed the name to make it more clear.